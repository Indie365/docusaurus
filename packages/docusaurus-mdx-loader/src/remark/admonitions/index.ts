/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import visit from 'unist-util-visit';
// @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
import type {Transformer, Processor, Plugin} from 'unified';

// @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
import type {ContainerDirective} from 'mdast-util-directive';
import type {Node} from 'unist';
import type {Parent} from 'mdast';

// TODO not ideal option shape
// First let upgrade to MDX 2.0
// Maybe we'll want to provide different tags for different admonition types?
// Also maybe rename "keywords" to "types"?
export type AdmonitionOptions = {
  tag: string; // TODO remove this config option, now unused with md directives
  keywords: string[];
  extendDefaults: boolean;
};

export const DefaultAdmonitionOptions: AdmonitionOptions = {
  tag: ':::', // TODO remove this, breaking change
  keywords: [
    'secondary',
    'info',
    'success',
    'danger',
    'note',
    'tip',
    'warning',
    'important',
    'caution',
  ],
  extendDefaults: false, // TODO make it true by default: breaking change
};

export function normalizeAdmonitionOptions(
  providedOptions: Partial<AdmonitionOptions> | true,
): AdmonitionOptions {
  if (providedOptions === true) {
    return DefaultAdmonitionOptions;
  }

  const options = {...DefaultAdmonitionOptions, ...providedOptions};

  // By default it makes more sense to append keywords to the default ones
  // Adding custom keywords is more common than disabling existing ones
  if (options.extendDefaults) {
    options.keywords = [
      ...DefaultAdmonitionOptions.keywords,
      ...options.keywords,
    ];
  }

  return options;
}

/*
const isMdxEsmLiteral = (node: Node): node is Literal =>
  node.type === 'mdxjsEsm';
// TODO good-enough approximation, but not 100% accurate
const isAdmonitionImport = (node: Node): boolean =>
  isMdxEsmLiteral(node) && node.value.includes('@theme/Admonition');

function createImportNode() {
  return {
    type: 'mdxjsEsm',
    value: "import Admonition from '@theme/Admonition'",
    data: {
      estree: {
        type: 'Program',
        body: [
          {
            type: 'ImportDeclaration',
            specifiers: [
              {
                type: 'ImportDefaultSpecifier',
                local: {type: 'Identifier', name: 'Admonition'},
              },
            ],
            source: {
              type: 'Literal',
              value: '@theme/Admonition',
              raw: "'@theme/Admonition'",
            },
          },
        ],
        sourceType: 'module',
      },
    },
  };
}

function createAdmonitionNode({
  directive,
  title,
  contentNodes,
}: {
  directive: ContainerDirective;
  title: unknown;
  contentNodes: unknown;
}) {
  return {
    type: 'mdxJsxFlowElement',
    name: 'Admonition',
    attributes: [
      {
        type: 'mdxJsxAttribute',
        name: 'type',
        value: directive.name,
      },
      {
        type: 'mdxJsxAttribute',
        name: 'title',
        value: {
          type: 'mdxJsxAttributeValueExpression',
          // TODO this is the complex part I couldn't solve  :/
          value: mdastToValueExpression(title),
        },
      },
    ],
    children: contentNodes,
  };
}
 */

type DirectiveLabel = Parent;

function parseDirective(directive: ContainerDirective): {
  directiveLabel: DirectiveLabel | undefined;
  contentNodes: Node[];
} {
  const hasDirectiveLabel =
    directive.children?.[0].data?.directiveLabel === true;
  if (hasDirectiveLabel) {
    const [directiveLabel, ...contentNodes] = directive.children;
    return {directiveLabel, contentNodes};
  }
  return {directiveLabel: undefined, contentNodes: directive.children};
}

function getTextOnlyTitle(directiveLabel: DirectiveLabel): string | undefined {
  const isTextOnlyTitle =
    directiveLabel?.children?.length === 1 &&
    directiveLabel?.children?.[0]?.type === 'text';
  return isTextOnlyTitle
    ? // @ts-expect-error: todo type
      (directiveLabel?.children?.[0].value as string)
    : undefined;
}

// This string value does not matter much
// It is ignored because nodes are using hName/hProperties coming from HAST
// const admonitionNodeType = 'admonitionHTML';

const plugin: Plugin = function plugin(
  this: Processor,
  optionsInput: Partial<AdmonitionOptions> = {},
): Transformer {
  const {keywords} = normalizeAdmonitionOptions(optionsInput);

  return async (root) => {
    /*
    let transformed = false;
    let alreadyImported = false;
     */

    visit(root, (node) => {
      /*
      if (isAdmonitionImport(node)) {
        alreadyImported = true;
      }
       */

      // See also:
      // https://talk.commonmark.org/t/generic-directives-plugins-syntax/444
      // https://github.com/remarkjs/remark-directive
      if (node.type === 'containerDirective') {
        // transformed = true;

        const directive = node as ContainerDirective;
        const isAdmonition = keywords.includes(directive.name);

        if (!isAdmonition) {
          return;
        }

        const {directiveLabel, contentNodes} = parseDirective(directive);

        /*
        // :::tip{title="my title} overrides :::tip[my title]
        const title = directive.attributes?.title ?? directiveLabel?.children;
        const admonitionNode = createAdmonitionNode({
          directive,
          title,
          contentNodes,
        });
        parent!.children.splice(index, 1, admonitionNode);
         */

        const textOnlyTitle =
          directive.attributes?.title ??
          (directiveLabel ? getTextOnlyTitle(directiveLabel) : undefined);

        // Transform the mdast directive node to a hast admonition node
        // See https://github.com/syntax-tree/mdast-util-to-hast#fields-on-nodes
        // TODO in MDX v2 we should transform the whole directive to
        // mdxJsxFlowElement instead of using hast
        directive.data = {
          hName: 'admonition',
          hProperties: {
            ...(textOnlyTitle && {title: textOnlyTitle}),
            type: directive.name,
          },
        };
        directive.children = contentNodes;

        // TODO legacy MDX v1 <mdxAdmonitionTitle> workaround
        // v1: not possible to inject complex JSX elements as props
        // v2: now possible: use a mdxJsxFlowElement element
        if (directiveLabel && !textOnlyTitle) {
          const complexTitleNode = {
            type: 'mdxAdmonitionTitle',
            data: {
              hName: 'mdxAdmonitionTitle',
              hProperties: {},
            },
            children: directiveLabel.children,
          };
          directive.children.unshift(complexTitleNode);
        }
      }
    });

    /*
    if (transformed && !alreadyImported) {
      (root as Parent).children.unshift(createImportNode());
    }
     */
  };
};

export default plugin;
