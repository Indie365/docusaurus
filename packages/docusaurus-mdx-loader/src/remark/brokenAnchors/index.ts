/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import logger from '@docusaurus/logger';
import type {Heading, Link, Text} from 'mdast';
// @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
import type {Transformer, Processor, Parent} from 'unified';

// TODO as of April 2023, no way to import/re-export this ESM type easily :/
// This might change soon, likely after TS 5.2
// See https://github.com/microsoft/TypeScript/issues/49721#issuecomment-1517839391
// import type {Plugin} from 'unified';
type Plugin = any; // TODO fix this asap

const nodeTypes: ['heading', 'link'] = ['heading', 'link'];

type AnchorList = {
  links: {link: {url?: string; anchor?: string}}[];
  headings: {heading: {text: string}}[];
};

type NodeAnchorList = {
  nodes: AnchorList;
  filePath: string;
};

function isValidURL(url: string): boolean {
  try {
    // eslint-disable-next-line no-new
    new URL(url);
    return true;
  } catch {
    return false;
  }
}

function checkAnchor(anchor: NodeAnchorList) {
  const {links, headings} = anchor.nodes;

  const invalidAnchor: string[] = [];

  const headingsText = headings.map((heading) =>
    heading.heading.text.toLowerCase(),
  );

  // filter out the anchor that has a url (pointing to another .md file)
  const linkAnchor = links.filter((link) => link.link.url === undefined);

  linkAnchor.forEach((el) => {
    // if the anchor is undefined, then skip
    if (
      el.link.anchor &&
      !headingsText.includes(el.link.anchor.toLowerCase())
    ) {
      invalidAnchor.push(el.link.anchor);
    }
  });

  if (invalidAnchor.length > 0) {
    const anchorLength = logger.interpolate`number=${invalidAnchor.length}`;
    const filePath = logger.interpolate`path=${anchor.filePath}`;
    const invalidAnchors = invalidAnchor.join(', ');
    logger.warn`Docusaurus found ${anchorLength} broken anchor(s) in file ${filePath}:\n${invalidAnchors}`;
  }
}

const plugin: Plugin = function plugin(this: Processor): Transformer {
  return async (tree, file) => {
    const {visit} = await import('unist-util-visit');

    const anchorList: AnchorList = {links: [], headings: []};
    visit<Parent, ['heading', 'link']>(
      tree,
      nodeTypes,
      (directive: Heading | Link) => {
        // we want only markdown links
        if (directive.type === 'link' && isValidURL(directive?.url)) {
          return;
        }

        // if the heading is empty then skip
        if (directive.type === 'heading' && directive.children.length !== 0) {
          // if the heading is a link, then check the anchor
          const childNode = directive.children[0] as Link | Text;
          if (childNode?.type === 'link' && childNode.children.length !== 0) {
            const linkTextValue = childNode.children[0] as any;
            anchorList.headings.push({
              heading: {text: linkTextValue.value.replaceAll(' ', '-')},
            });
          } else if (childNode?.type === 'text') {
            anchorList.headings.push({
              heading: {
                text: childNode.value.replaceAll(' ', '-'),
              },
            });
          }
        }

        // get the URL of link nodes
        if (directive.type === 'link') {
          // check if link isn't empty
          // ? should we report empty links?
          if (directive.children.length !== 0) {
            const linkUrlValue = directive.url;
            const [url, anchor] = linkUrlValue.split('#');
            anchorList.links.push({
              link: {
                url: url === '' ? undefined : url,
                anchor: anchor === '' ? undefined : anchor,
              },
            });
          }
        }
      },
    );

    // We only enable these warnings for the client compiler
    // This avoids emitting duplicate warnings in prod mode
    // Note: the client compiler is used in both dev/prod modes
    if (file.data.compilerName === 'client') {
      checkAnchor({
        nodes: anchorList,
        filePath: file.path,
      });
    }
  };
};

export default plugin;
