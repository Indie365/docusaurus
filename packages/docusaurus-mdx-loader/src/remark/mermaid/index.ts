/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import visit from 'unist-util-visit';
import type {Transformer} from 'unified';
import type {Data, Literal, Node, Parent} from 'unist';
import type {Code} from 'mdast';

function processMermaidNode(
  node: Code,
  index: number,
  parent: Parent<Node<Data> | Literal, Data>,
) {
  parent.children.splice(index, 1, {
    type: 'mermaidCodeBlock',
    data: {
      hName: 'mermaid',
      hProperties: {
        value: node.value,
      },
    },
  });
}

export default function plugin(): Transformer {
  return (root) => {
    visit(root, 'code', (node: Code, index, parent) => {
      if (node.lang === 'mermaid') {
        processMermaidNode(node, index, parent!);
      }
    });
  };
}
