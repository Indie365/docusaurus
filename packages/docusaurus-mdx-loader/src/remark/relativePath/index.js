/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

const visit = require('unist-util-visit');
const path = require('path');
const url = require('url');
const fs = require('fs-extra');

const plugin = (options) => {
  const transformer = (root) => {
    visit(root, 'image', (node) => {
      if (!url.parse(node.url).protocol) {
        if (!path.isAbsolute(node.url)) {
          node.type = 'jsx';
          node.value = `<img ${node.alt ? `alt={"${node.alt}"}` : ''} ${
            node.url
              ? `src={require("!url-loader!${
                  node.url.startsWith('./') ? node.url : `./${node.url}`
                }").default}`
              : ''
          } ${node.title ? `title={"${node.title}"}` : ''} />`;
          if (node.url) {
            delete node.url;
          }
          if (node.alt) {
            delete node.alt;
          }
          if (node.title) {
            delete node.title;
          }
        } else if (!fs.existsSync(path.join(options.staticDir, node.url))) {
          throw new Error(`File ${node.url} not found in ${options.staticDir}`);
        }
      }
    });
  };

  return transformer;
};

module.exports = plugin;
