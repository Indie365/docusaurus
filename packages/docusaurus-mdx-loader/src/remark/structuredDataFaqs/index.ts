/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {parse, type ParserOptions} from '@babel/parser';
import traverse from '@babel/traverse';
import stringifyObject from 'stringify-object';
import toString from 'mdast-util-to-string';
import visit from 'unist-util-visit';
import {toMessageRelativeFilePath} from '@docusaurus/utils';
import {toValue} from '../utils';

import type {Identifier} from '@babel/types';
import type {Node, Parent} from 'unist';
import type {Heading, Literal, Paragraph} from 'mdast';
import type {Transformer} from 'unified';

const parseOptions: ParserOptions = {
  plugins: ['jsx'],
  sourceType: 'module',
};

const name = 'structuredDataFaqs';

const isImport = (child: Node): child is Literal => child.type === 'import';
const hasImports = (index: number) => index > -1;
const isExport = (child: Node): child is Literal => child.type === 'export';

const isTarget = (child: Literal) => {
  let found = false;
  const ast = parse(child.value, parseOptions);
  traverse(ast, {
    VariableDeclarator: (path) => {
      if ((path.node.id as Identifier).name === name) {
        found = true;
      }
    },
  });
  return found;
};

const getOrCreateExistingTargetIndex = (children: Node[]) => {
  let importsIndex = -1;
  let targetIndex = -1;

  children.forEach((child, index) => {
    if (isImport(child)) {
      importsIndex = index;
    } else if (isExport(child) && isTarget(child)) {
      targetIndex = index;
    }
  });

  if (targetIndex === -1) {
    const target = {
      default: false,
      type: 'export',
      value: `export const ${name} = [];`,
    };

    targetIndex = hasImports(importsIndex) ? importsIndex + 1 : 0;
    children.splice(targetIndex, 0, target);
  }

  return targetIndex;
};

export default function plugin(): Transformer {
  return (root, vfile) => {
    let rootFaqsHeading: Heading | undefined;
    let readingFAQs = false;
    const faqsNodes: {heading: Heading; paragraphs: Paragraph[]}[] = []; // faqs question -> faqs answer
    let x = 0;
    visit(
      root,
      ['heading', 'paragraph'],
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      (child: Heading | Paragraph, index, parent) => {
        if (vfile.path?.includes('blog/2022-09-01-docusaurus-2.1/index.mdx')) {
          x += 1;
          console.log(x, 'child', vfile.path, child.type);
        }

        if (readingFAQs && child.type === 'paragraph' && faqsNodes.length > 0) {
          // faqs answers are paragraphs
          faqsNodes[faqsNodes.length - 1]!.paragraphs.push(child);
        } else if (child.type === 'heading') {
          // faqs questions are headings

          const value = toString(child);
          if (child.depth === 2 && value === 'FAQs') {
            if (rootFaqsHeading) {
              throw new Error(
                `Found multiple FAQs headings in ${toMessageRelativeFilePath(
                  vfile.path!,
                )}. Please ensure you only have one faqs heading.`,
              );
            }

            // this is the root faqs heading
            rootFaqsHeading = child;
            readingFAQs = true;
            return;
          }

          if (child.depth === 2 && readingFAQs) {
            // this is the end of the faqs section
            readingFAQs = false;
            return;
          }

          // const x = parent === root;
          // console.log(x)

          if (child.depth === 3 && readingFAQs) {
            // this is a faqs question
            faqsNodes.push({
              heading: child,
              paragraphs: [],
            });
          }

          // // depth: 1 headings are titles and not included in the TOC
          // if (parent !== root || !value || child.depth < 2) {
          //   return;
          // }

          // headings.push({
          //   value: toValue(child),
          //   id: child.data!.id as string,
          //   level: child.depth,
          // });
        }
      },
    );
    const {children} = root as Parent<Literal>;
    const targetIndex = getOrCreateExistingTargetIndex(children);

    if (faqsNodes.length > 0) {
      const faqs = faqsNodes.map(({heading, paragraphs}) => ({
        question: toValue(heading),
        answer: paragraphs.map(toString).join(' '),
      }));

      if (vfile.path?.includes('blog/2022-09-01-docusaurus-2.1/index.mdx')) {
        console.log(x, 'faqsNodes', vfile.path, faqsNodes, faqs, root);
        console.log('faqsNodes', vfile.path, faqs);
      }
      children[targetIndex]!.value = `export const ${name} = ${stringifyObject(
        faqs,
      )};`;
    }
  };
}
