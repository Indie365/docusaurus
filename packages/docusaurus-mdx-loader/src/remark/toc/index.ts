/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {parse, type ParserOptions} from '@babel/parser';
import traverse from '@babel/traverse';
import {generate} from 'astring';
import {toValue} from '../utils';
import {hasImports, isExport, isImport} from './utils';
import type {TOCItem, NestedTOC} from './utils';
import type {
  SpreadElement,
  Program,
  ImportDeclaration,
  ImportSpecifier,
} from 'estree';
import type {Identifier} from '@babel/types';
import type {Node, Parent} from 'unist';
import type {Heading, Literal} from 'mdast';
// @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
import type {Transformer} from 'unified';
import type {
  MdxjsEsm,
  MdxJsxFlowElement,
  // @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
} from 'mdast-util-mdx';

// Reexport TOCItem, since it's used throughout the project
export type {TOCItem} from './utils';

// TODO as of April 2023, no way to import/re-export this ESM type easily :/
// TODO upgrade to TS 5.3
// See https://github.com/microsoft/TypeScript/issues/49721#issuecomment-1517839391
// import type {Plugin} from 'unified';
type Plugin = any;

const parseOptions: ParserOptions = {
  plugins: ['jsx'],
  sourceType: 'module',
};

interface PluginOptions {
  name?: string;
}

const isTarget = (child: Literal, name: string) => {
  let found = false;
  const ast = parse(child.value, parseOptions);
  traverse(ast, {
    VariableDeclarator: (path) => {
      if ((path.node.id as Identifier).name === name) {
        found = true;
      }
    },
  });
  return found;
};

const getOrCreateExistingTargetIndex = async (
  children: Node[],
  name: string,
) => {
  let importsIndex = -1;
  let targetIndex = -1;

  children.forEach((child, index) => {
    if (isImport(child)) {
      importsIndex = index;
    } else if (isExport(child) && isTarget(child, name)) {
      targetIndex = index;
    }
  });

  if (targetIndex === -1) {
    const target = await createExportNode(name, [], []);

    targetIndex = hasImports(importsIndex) ? importsIndex + 1 : 0;
    children.splice(targetIndex, 0, target);
  }

  return targetIndex;
};

const plugin: Plugin = function plugin(
  options: PluginOptions = {},
): Transformer {
  const name = options.name || 'toc';

  return async (root) => {
    const {toString} = await import('mdast-util-to-string');
    const {visit} = await import('unist-util-visit');

    const partialComponentToHeadingsName: {[key: string]: string} =
      Object.create(null);

    const headings: (TOCItem | NestedTOC)[] = [];
    const imports: ImportDeclaration[] = [];

    function visitHeading(node: Heading) {
      const value = toString(node);

      // depth:1 headings are titles and not included in the TOC
      if (!value || node.depth < 2) {
        return;
      }

      headings.push({
        value: toValue(node, toString),
        id: node.data!.id!,
        level: node.depth,
      });
    }

    function visitMdxjsEsm(node: MdxjsEsm) {
      if (!node.data?.estree) {
        return;
      }

      for (const potentialImportDeclaration of node.data.estree.body) {
        if (potentialImportDeclaration.type !== 'ImportDeclaration') {
          continue;
        }

        const importPath = potentialImportDeclaration.source.value as string;
        const isMdxImport = /\.mdx?$/.test(importPath);
        if (!isMdxImport) {
          continue;
        }

        const componentName = potentialImportDeclaration.specifiers.find(
          (o: Node) => o.type === 'ImportDefaultSpecifier',
        )?.local.name;

        if (!componentName) {
          continue;
        }
        const {length} = Object.keys(partialComponentToHeadingsName);
        const exportAsName = `${name}${length}`;
        partialComponentToHeadingsName[componentName] = exportAsName;

        const specifier: ImportSpecifier = {
          type: 'ImportSpecifier',
          imported: {type: 'Identifier', name},
          local: {type: 'Identifier', name: exportAsName},
        };

        imports.push({
          type: 'ImportDeclaration',
          specifiers: [specifier],
          source: potentialImportDeclaration.source,
        });
        potentialImportDeclaration.specifiers.push(specifier);
      }
    }

    function visitMdxJsxFlowElement(node: MdxJsxFlowElement) {
      const nodeName = node.name;
      if (!nodeName) {
        return;
      }
      const headingsName = partialComponentToHeadingsName[nodeName];
      if (headingsName) {
        headings.push({
          nested: true,
          name: headingsName,
        });
      }
    }

    visit(root, ['heading', 'mdxjsEsm', 'mdxJsxFlowElement'], (child) => {
      if (child.type === 'heading') {
        visitHeading(child as Heading);
      } else if (child.type === 'mdxjsEsm') {
        visitMdxjsEsm(child as MdxjsEsm);
      } else if (child.type === 'mdxJsxFlowElement') {
        visitMdxJsxFlowElement(child as MdxJsxFlowElement);
      }
    });

    const {children} = root as Parent;
    const targetIndex = await getOrCreateExistingTargetIndex(children, name);

    if (headings?.length) {
      children[targetIndex] = await createExportNode(name, headings, imports);
    }
  };
};

export default plugin;

async function createExportNode(
  name: string,
  headings: (TOCItem | NestedTOC)[],
  imports: ImportDeclaration[],
): Promise<MdxjsEsm> {
  const {valueToEstree} = await import('estree-util-value-to-estree');

  const tocObject = headings.map((heading) => {
    if ('nested' in heading) {
      const spreadElement: SpreadElement = {
        type: 'SpreadElement',
        argument: {type: 'Identifier', name: heading.name},
      };
      return spreadElement;
    }

    return valueToEstree(heading);
  });

  const estree: Program = {
    type: 'Program',
    body: [
      ...imports,
      {
        type: 'ExportNamedDeclaration',
        declaration: {
          type: 'VariableDeclaration',
          declarations: [
            {
              type: 'VariableDeclarator',
              id: {
                type: 'Identifier',
                name,
              },
              init: {
                type: 'ArrayExpression',
                elements: tocObject,
              },
            },
          ],
          kind: 'const',
        },
        specifiers: [],
        source: null,
      },
    ],
    sourceType: 'module',
  };

  return {
    type: 'mdxjsEsm',
    value: generate(estree),
    data: {
      estree,
    },
  };
}
