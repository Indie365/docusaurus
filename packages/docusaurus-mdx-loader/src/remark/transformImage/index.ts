/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import visit, {Visitor} from 'unist-util-visit';
import path from 'path';
import url from 'url';
import fs from 'fs-extra';
import escapeHtml from 'escape-html';
import {getFileLoaderUtils} from '@docusaurus/core/lib/webpack/utils';
import {
  posixPath,
  escapePath,
  toMessageRelativeFilePath,
} from '@docusaurus/utils';
import type {Plugin, Transformer} from 'unified';
import type {Image} from 'mdast';

const {
  loaders: {inlineMarkdownImageFileLoader},
} = getFileLoaderUtils();

interface PluginOptions {
  filePath: string;
  staticDir: string;
}

const createJSX = (node: Image, pathUrl: string) => {
  const {url: _url, alt, title, ...strippedNode} = node;
  const jsxNode = {
    ...strippedNode,
    type: 'jsx',
    value: `<img ${node.alt ? `alt={"${escapeHtml(node.alt)}"} ` : ''}${
      node.url
        ? `src={require("${inlineMarkdownImageFileLoader}${escapePath(
            pathUrl,
          )}").default}`
        : ''
    }${node.title ? ` title="${escapeHtml(node.title)}"` : ''} />`,
  };
  return jsxNode;
};

async function ensureImageFileExist(imagePath: string, sourceFilePath: string) {
  const imageExists = await fs.pathExists(imagePath);
  if (!imageExists) {
    throw new Error(
      `Image ${toMessageRelativeFilePath(
        imagePath,
      )} used in ${toMessageRelativeFilePath(sourceFilePath)} not found.`,
    );
  }
}

async function processImageNode(
  node: Image,
  {filePath, staticDir}: PluginOptions,
) {
  if (!node.url) {
    throw new Error(
      `Markdown image URL is mandatory in "${toMessageRelativeFilePath(
        filePath,
      )}" file`,
    );
  }

  const parsedUrl = url.parse(node.url);
  if (parsedUrl.protocol) {
    // pathname:// is an escape hatch,
    // in case user does not want his images to be converted to require calls going through webpack loader
    // we don't have to document this for now,
    // it's mostly to make next release less risky (2.0.0-alpha.59)
    if (parsedUrl.protocol === 'pathname:') {
      node.url = node.url.replace('pathname://', '');
    } else {
      // noop
    }
  }
  // images without protocol
  else if (path.isAbsolute(node.url)) {
    // absolute paths are expected to exist in the static folder
    const expectedImagePath = path.join(staticDir, node.url);
    await ensureImageFileExist(expectedImagePath, filePath);
    createJSX(node, posixPath(expectedImagePath));
  }
  // We try to convert image urls without protocol to images with require calls
  // going through webpack ensures that image assets exist at build time
  else {
    // relative paths are resolved against the source file's folder
    const expectedImagePath = path.join(path.dirname(filePath), node.url);
    await ensureImageFileExist(expectedImagePath, filePath);
    createJSX(node, node.url.startsWith('./') ? node.url : `./${node.url}`);
  }
}

const plugin: Plugin<[PluginOptions]> = (options) => {
  const transformer: Transformer = async (root) => {
    const promises: Promise<void>[] = [];
    const visitor: Visitor<Image> = (node) => {
      promises.push(processImageNode(node, options));
    };
    visit(root, 'image', visitor);
    await Promise.all(promises);
  };
  return transformer;
};

export default plugin;
