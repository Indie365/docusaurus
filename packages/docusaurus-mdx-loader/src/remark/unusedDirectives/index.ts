/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import path from 'path';
import process from 'process';
import visit from 'unist-util-visit';
import logger from '@docusaurus/logger';
import {posixPath} from '@docusaurus/utils';
// @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
import type {Transformer, Processor, Parent} from 'unified';
import type {
  Directive,
  // @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
} from 'mdast-util-directive';

// TODO as of April 2023, no way to import/re-export this ESM type easily :/
// This might change soon, likely after TS 5.2
// See https://github.com/microsoft/TypeScript/issues/49721#issuecomment-1517839391
// import type {Plugin} from 'unified';
type Plugin = any; // TODO fix this asap

const directiveTypes = ['containerDirective', 'leafDirective', 'textDirective'];

const plugin: Plugin = function plugin(this: Processor): Transformer {
  return (tree, file) => {
    // We only enable these warnings for the client compiler
    // This avoids emitting duplicate warnings in prod mode
    // Note: the client compiler is used in both dev/prod modes
    if (file.data.compilerName !== 'client') {
      return;
    }

    const unusedDirectives: Array<{
      name: string;
      type: string;
      position:
        | {
            line: number;
            column: number;
          }
        | undefined;
    }> = [];

    visit<Parent>(tree, directiveTypes, (node: Directive) => {
      if (!node.data) {
        unusedDirectives.push({
          name: node.name,
          type: node.type,
          position: node.position
            ? {
                line: node.position.start.line,
                column: node.position.start.column,
              }
            : undefined,
        });
      }
    });

    if (unusedDirectives.length > 0) {
      const warningMessage = unusedDirectives
        .map((unusedDirective) => {
          const positionMessage = unusedDirective.position
            ? logger.interpolate`number=${unusedDirective.position.line}:number=${unusedDirective.position.column}`
            : '';

          const customPath = posixPath(path.relative(process.cwd(), file.path));

          return logger.interpolate`We found a potential error in your documentation name=${
            unusedDirective.name
          } path=${`${customPath}:${positionMessage}`}`;
        })
        .join('\n');
      logger.warn(warningMessage);
    }
  };
};

export default plugin;
