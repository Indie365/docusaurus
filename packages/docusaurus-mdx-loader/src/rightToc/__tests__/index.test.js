/**
 * Copyright (c) 2017-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import '@babel/polyfill';
import {join} from 'path';
import remark from 'remark';
import mdx from 'remark-mdx';
import vfile from 'to-vfile';
import plugin from '../index';

const processFixture = async (name, options) => {
  const path = join(__dirname, 'fixtures', `${name}.mdx`);
  const file = await vfile.read(path);
  const result = await remark()
    .use(mdx)
    .use(plugin, options)
    .process(file);

  return result.toString();
};

test('no options', async () => {
  const options = undefined;
  const result = await processFixture('just-content', options);
  expect(result).toMatchSnapshot();
});

test('should export even with existing name', async () => {
  const result = await processFixture('name-exist');
  expect(result).toMatchSnapshot();
});

test('should export with custom name', async () => {
  const options = {
    name: 'customName',
  };
  const result = await processFixture('just-content', options);
  expect(result).toMatchSnapshot();
});

test('should insert below imports', async () => {
  const result = await processFixture('insert-below-imports');
  expect(result).toMatchSnapshot();
});

test('empty headings', async () => {
  const result = await processFixture('empty-headings');
  expect(result).toMatchSnapshot();
});
