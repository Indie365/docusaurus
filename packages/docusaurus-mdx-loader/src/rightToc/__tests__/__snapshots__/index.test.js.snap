// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty headings 1`] = `
"export const rightToc = [];

# Ignore this

## 

## ![](an-image.svg)
"
`;

exports[`no options 1`] = `
"export const rightToc = [
	{
		value: 'Endi',
		id: 'endi',
		children: []
	},
	{
		value: 'Endi',
		id: 'endi-1',
		children: [
			{
				value: 'Yangshun',
				id: 'yangshun',
				children: []
			}
		]
	},
	{
		value: 'I ♥ unicode.',
		id: 'i--unicode',
		children: []
	}
];

### Endi

\`\`\`md
## This is ignored
\`\`\`

## Endi

Lorem ipsum

### Yangshun

Some content here

## I ♥ unicode.
"
`;

exports[`should export even with existing name 1`] = `
"export const rightToc = [
	{
		value: 'Thanos',
		id: 'thanos',
		children: []
	},
	{
		value: 'Tony Stark',
		id: 'tony-stark',
		children: [
			{
				value: 'Avengers',
				id: 'avengers',
				children: []
			}
		]
	}
];

## Thanos

## Tony Stark

### Avengers
"
`;

exports[`should export with custom name 1`] = `
"export const customName = [
	{
		value: 'Endi',
		id: 'endi',
		children: []
	},
	{
		value: 'Endi',
		id: 'endi-1',
		children: [
			{
				value: 'Yangshun',
				id: 'yangshun',
				children: []
			}
		]
	},
	{
		value: 'I ♥ unicode.',
		id: 'i--unicode',
		children: []
	}
];

### Endi

\`\`\`md
## This is ignored
\`\`\`

## Endi

Lorem ipsum

### Yangshun

Some content here

## I ♥ unicode.
"
`;

exports[`should insert below imports 1`] = `
"import something from 'something';

import somethingElse from 'something-else';

export const rightToc = [
	{
		value: 'Title',
		id: 'title',
		children: []
	},
	{
		value: 'Test',
		id: 'test',
		children: [
			{
				value: 'Again',
				id: 'again',
				children: []
			}
		]
	}
];

## Title

## Test

### Again

Content.
"
`;
